### PR TITLE
Support for the customer reference tagging

### LINKED ISSUE #32: Support for customer reference

### PR DESCRIPTION
**Pull request by mridang**
Closes #32

### CODE REVIEW

File: CustomerData/CartTagging.php [hunk 1]
```diff
@@ -181,6 +187,19 @@ private function updateNostoId()
     }
 
     /**
+     * Return the checksum for for the customer tagging i.e hashed cookie identifier or HCID for
+     * short. This is used to sign the tagging so that if it is in fact cached, the cookie and
+     * tagging signature won't match and we'll be able to show a warning.
```
Comment (oostman) [inline CustomerData/CartTagging.php]: Do we show and warning? 
I would say 'and we'll be able to discard the cached data'

File: CustomerData/CustomerTagging.php [hunk 2]
```diff
@@ -76,10 +80,36 @@ public function getSectionData()
...
+    /**
+     * Return the checksum / customer reference for customer
+     *
+     * @param CustomerInterface $customer
+     * @return string
+     */
+    public function generateCustomerReference(CustomerInterface $customer)
+    {
+        /** @noinspection PhpUndefinedMethodInspection */
+        return uniqid(substr(md5($customer->getId() . $customer->getEmail()), 0, 8), true);
```
Comment (oostman) [inline CustomerData/CustomerTagging.php]: Why use md5? I would use sha256 here as well, unless the performance impact is bad.
What does uniqid do, does it guarantee uniqueness?

File: Helper/Data.php [hunk 3]
```diff
@@ -96,18 +92,6 @@ public function __construct(
...
-     * @param string $string
-     *
-     * @return string
-     */
-    public static function generateVisitorChecksum($string)
-    {
-        return hash(self::VISITOR_HASH_ALGO, $string);
-    }
-
-    /**
```
Comment (hannunosto) [inline Helper/Data.php]: Why was this moved from here? Now we have the method duplicated in https://github.com/Nosto/nosto-magento2/blob/feature/customer-reference-support/CustomerData/CartTagging.php#L197 and
https://github.com/Nosto/nosto-magento2/blob/feature/customer-reference-support/CustomerData/CustomerTagging.php#L99. I would keep it in one place

File: CustomerData/CustomerTagging.php (hunk 2 above)
Comment (hannunosto) [inline CustomerData/CustomerTagging.php]: We cannot use `uniqid()` unless we save the customer reference. `uniqid()` creates unique number. The first attribute is a prefix. In this case the prefix would be `substr(md5($customer->getId() . $customer->getEmail()), 0, 8)`.

Comment (mridang) [inline CustomerData/CustomerTagging.php]: Honestly, this code has been largely moved from M1 and has been tried and battle tested. ;)

File: Helper/Data.php (hunk 3 above)
Comment (mridang) [inline Helper/Data.php]: I created base class and now the private-data tagging is in one place

File: CustomerData/CustomerTagging.php (hunk 2 above)
Comment (mridang) [inline CustomerData/CustomerTagging.php]: I removed it.
